### DISCUSSION TITLE
link to an alias in another mindmap?

### DISCUSSION THREAD

**Discussion by wordmuse** [freeplane/freeplane#268] [Ideas]
Within a single mindmap, I can have a link that looks like this: #at(:~aliasname)

It would be nice if I could reference the alias from within a 2nd mindmap. 

It would be even sweeter if I could create a hyperlink from, say Word, that did this, too.

Is there currently a way to do this? If not, could it be considered?

~ Bal

**Comment by nnako** [comment]
Hi @wordmuse ,

interesting. I didn't even notice there were new ways to link within a map. Now saw that your mention was introduced in `v1.7.x`. unfortunately I can't find a documentation for this new feature.

As far as I know, linking within maps can happen in the following ways:

**INSIDE ONE MAP**

- optical arrow links e.g. by draging between two nodes with `<RIGHT-MOUSE>` pressed
- textual links e.g. to other nodes by using `=ID_12345678.text` within the node's core. the thus modified node by default will be surrounded by a green box to indicate that it`s core is automatically synchronized.
- hyperlinks e.g. by selecting two nodes and pressing `<SHIFT> + <ALT> + <L>` which creates a hyperlink entry within the 1st node's link attribute like `#ID_12345678`

**BETWEEN MAPS**

- when using hyperlinks and add a mindmap path to the front of it, links between the nodes of different mindmaps can be established. like `file:<ABSOLUTE-PATH-TO-MM-FILE>#ID_12345678` within the source node's link attribute

Now, please, could you explain (or link me with the source of) the format specification for the feature you mentioned: `#at(:~aliasname)`? where would the `aliasname` have to be defined in the first place? is the `at` a fix token or does it rather represent something like a target node id? which other format variations are valid when varying `#`, `:`, `(` and `) `?

Thanks.

**Comment by wordmuse** [comment]
Hi nnako,

Check this out.
https://sourceforge.net/p/freeplane/discussion/758437/thread/f84a5fd554/

I can't explain _why_ "#at(:~aliasname) is the syntax for linking to an alias. You'll have to take that up with the developers. But the description of what can be done can be found in the discussion at the above link.

I use aliases a lot in my mindmaps. The ability to link to an alias within a second mindmap seems like a useful addition.

~ Bal

**Reply by nnako** [reply]
Hi @wordmuse , thanks. understood.

Have to say that the `alias` feature doesn't seems to have a good usability, yet: to establish a link (even **within** the same mindmap), you would have to remember the formerly given alias name and even type it into the dialog window. Why don't you use the `anchor` feature?

The `anchor` feature works regardless of to where you want to establish the link. Just within the local mindmap or to a node within any other mindmap. The only "drawback" would be that there is no specific alias name per target to be used. But the target node is only temporarily valid.

Please try this and decide if the `anchor` feature would help you:

1. open both mindmaps, the one holding the source node and the one holding the target node within the same Freeplane application instance. 
2. select your target node within the target mindmap
3. select the menu item `insert` > `hyperlink` > `set anchor` . with this step, Freeplane will remember the selected node as a usable reference until it is set to another node. the node is now set as `anchor`.
4. switch to your source mindmap and select your source node
5. select the menu item `insert` > `hyperlink` > `create link to anchor` . this will establish a hyperlink from your currently selected source node to the previously selected `anchor` node (residing in the other mindmap).
6. check the hyperlink by clicking on the newly established link symbol (red). this connection will work as hyperlink even when the target mindmap is closed.

You could now, as the previously set `anchor` is still valid, connect other nodes from other currently opened mindmaps. And each time, you can decide whether to connect **from** or **to** the `anchor`.

To make things easier, I would recommend to set keyboard shortcuts for these three `anchor` functinos. here, my examples:

- `<CTRL> + <ALT> + <A>` to set the anchor (to the selected node within any open mindmap)
- `<CTRL> + <ALT> + <F>` to create a hyperlink FROM the anchor (set link within anchor node)
- `<CTRL> + <ALT> + <T>` to create a hyperlink TO the anchor (set link within local node)

Please have a look at the local node where the hyperlink has been established. It will show the format: `file:<ABSOLUTE-PATH-TO-MM-FILE>#ID_12345678`

What do you think? What would be your objections to using the `anchor` feature to create internal or external links?

**Reply by wordmuse** [reply]
I use aliases much like I use mnemonics. I can use the GoTo command, and in the dialog, enter something like **_:~aliasname_**, and go to the proper node from wherever I am. 

In my current request, what I'm envisioning is being able to simply remember the alias in the destination mindmap, and when I link to that mindmap, being able to additionally specify the alias, instead of having to travel through the mindmap to find the node that I want to link to.  It's like being able to say to someone, "Go to the 7-11 in Baxterville" instead of "Go  to 14323 Myrtle Avenue in Baxterville."

Also, I extensively use anchors, too. But they serve different needs. I don't need every node to be a named point in the mindmap. For those that are unnamed, anchors serve very well indeed. And I have yet to find a real use for visual links, as, aesthetically speaking, they seem to be more clutter than useful to me.

~ Bal

**Reply by nnako** [reply]
Hi Bal (@wordmuse), thanks for clarification. I understood :-) . And I will try to integrate your method into my own workflows, as I see great potential for optimized efficiency, here.

Concerning the usability of visual links or "arrow links", they seem to make sense when they are dynamically shown / removed in a local map. I use them in combination with scripts as to temporarily show e.g. which nodes within the map have a hyperlink to a specific alias node:

![grafik](https://user-images.githubusercontent.com/5845602/159223494-a3e362f3-ca47-4441-8152-bfcd7321784d.png)

I don't think, this would make any sense to show connections between nodes within **different** mindmaps, though.

**Comment by Illegit** [comment]
Hi wordmuse, have a look to this if it is helping:

https://sourceforge.net/p/freeplane/discussion/758437/thread/9c04bf8f99/?limit=25#4577
I utilize this structure as suggested there:
" =c.load('target map path').map.node('ID_xxxxxxx')['attribute name'].to.text "
SFPC

**Reply by Illegit** [reply]
and this:
c.load('prova1.mm').map.root.at(':~provaA')['progress']

SFPC

**Comment by macmarrum** [comment]
I know that node links and hyperlinks (within node core/details/note) which use a node ID work also between mindMaps.

#### Hyperlink example
Assuming both mindMaps are in the same directory, here's a sample code
```html
<html>
  <head>
    
  </head>
  <body>
    <p>
      This is a sentence in A.mm, linking to B.mm where you can 
      read more about <a href="B.mm#ID_1234567">sentences</a>.
    </p>
  </body>
</html>
```

**Comment by macmarrum** [comment]
An `alias` can be defined on a node in the right-click menu `Set node alias`, or in `Edit->Node properties->Set node alias`

`at()` syntax is described in the API docs for `NodeRO`

> Returns a single node located by path given as a string argument.
>
> <p>If no node or more than one node is available at the specified position,
> <code>NodeNotFoundException</code> is thrown.
>
> <p>The path is a concatenation of path elements described below.
>
>  <h2> Path examples:</h2>
>  <pre>
>  node.at(":'house'/'kitchen'/~chairs/~2")
>     - take global node with text 'house',
>     - in its subtree node find with text 'kitchen'
>     - find its child node with alias 'chairs'
>     - return the second child node of the chairs
>
>  node.at(".. 'house'/**&#47;~dog")
>      -- return node with alias 'dog' within subtree of ancestor node with text 'house'
>  </pre>
>
> Following path elements are defined:
>
> <h2>Child node:</h2>
> <pre>
> 'node text' or /"node text" : child node with text 'node text'
> 'node...' : child node containing text starting with 'node'
> * or '...' : any child node
> ~1 : child node at position 1 ( any number starting with 1 is allowed)
> ~nodeAlias : child node with alias 'nodeAlias', node alias can not be a number
> because numbers are reserved for the previous selector.
> </pre>
>
> <h2>Descendant node:</h2>
> <pre> /**&#47; </pre>
>
> <h2>Parent node:</h2>
> <pre> ..</pre>
>
> <h2>Ancestor node:</h2>
> <pre>
> ..'node text' or "node text" : the closest ancestor node with text 'node text'
> ..'node...' : the closest ancestor node containing text starting with 'node'
> ..~2  : second ancestor node also the parent node of the parent node (any positive number is allowed)
> ..~nodeAlias : the closest ancestor node with alias 'nodeAlias', node alias can not be a number
> </pre>
>
> <h2>Root node</h2>
> Prefix <b><code>/</code></b> as the first path character means the path starts from the root node.
>
> <pre>
> /'node text' or /"node text" : child node of the root with text 'node text'
> /'node...' : child node of the root containing text starting with 'node'
> /~nodeAlias : child node of the root with alias 'nodeAlias', node alias can not be a number
> </pre>
>
> <h2>Global node</h2>
> (allowed only as the first path element).
> Here global node is a node carrying global flag which can be set using menu or by script.
>
> <pre>
> :'node text' or :"node text" : global node with text 'node text'
> :'node...' : global node containing text starting with 'node'
> :~nodeAlias : global node with alias 'nodeAlias', node alias can not be a number
> </pre>

**Reply by nnako** [reply]
Hi @macmarrum , thanks for the hints. Looks like a powerful feature to be used within scripting, but maybe not so much during manual edit of maps. As of `alias`'s format complexity, I would definitely propose using the `anchor` feature, in Bal's case. Let's see what he thinks.

**Comment by macmarrum** [comment]
@nnako I can see why you would like to use other means to achieve the linking. And I can see why @wordmuse would like to use aliases. The concept of having `#at(:~some_alias)` as part of a link or a hyperlink makes absolute sense to me. I would +1 such a feature request.

Aliases are also useful to bookmark nodes. E.g. when I have several mindMaps with the same main structure and I want to jump easily to one of the main nodes, I can give them their aliases and use a script or a dialog box, like the one below, and not worry about creating anchors in each mindMap.
This is especially useful when each such a mindMap is created from a template, which already has the aliases set.

```groovy
// Go to alias
// @ExecutionModes({ON_SINGLE_NODE})
import org.freeplane.api.Node
import org.freeplane.features.explorer.NodeAlias
import org.freeplane.features.explorer.NodeAliases
import org.freeplane.features.map.NodeModel
import org.freeplane.features.mode.Controller
import org.freeplane.features.ui.IMapViewManager
import org.freeplane.plugin.script.proxy.ScriptUtils

import javax.swing.*
import java.awt.*

final c = ScriptUtils.c()
final node = ScriptUtils.node()
final NodeModel nodeModel = node.delegate

final Controller controller = Controller.getCurrentController()
final IMapViewManager viewController = controller.getMapViewManager()
final Component parentComponent = viewController.getComponent(nodeModel)

final Collection<NodeAlias> nodeAliases = NodeAliases.of(controller.map).aliases()
final String[] aliases = nodeAliases.collect { it.value }.sort().toArray()
String input = JOptionPane.showInputDialog(parentComponent, null, "Go to alias",
        JOptionPane.QUESTION_MESSAGE, null, aliases, null)
if (input == null || input == '') {
    c.statusInfo = 'input is null or blank'
    return
} else {
    Node dest = c.findAll().find { it.alias == input }
    if (dest)
        c.select(dest)
    else
        c.statusInfo = "$input not found"
}
```

**Comment by dpolivaev** [comment]
I think at all places where you can use references like `.../mindmap.mm#ID_1234567` you can also use references like `.../mindmap.mm#:'house'/'kitchen'/~chairs/~2`. If they are relative (do not start with a colon character `:`) they should be resolved relative to the root node. It should apply as for the links to mind maps as for Freeplane command line arguments.

**Comment by macmarrum** [comment]
@dpolivaev unfortunately it does not work.
Issue #269 created

**Comment by nnako** [comment]
@dpolivaev , the idea is great to have this format as an overall user interface to hyperlinks between local and "remote" nodes. but, no, it doesn't work, yet.

@macmarrum , I understand now, if you use template maps for the same kind of subjects and always want to have the opportunity to hyperlink to let's say `file:/<subject>.mm#:~<alias>` where all the alias names are fixed and globally known, there seems no way to do it without a (new) "global alias" feature.
